### PR TITLE
Add abseil/20200225.2

### PR DESCRIPTION
**Pull request by 0x8000-0000**
This is based on the latest tag (20200225.2) on the lts_2020_02_25 branch.

Specify library name and version:  **abseil/20200225.2**

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/abseil/all/conanfile.py [hunk 1]
```diff
@@ -147,5 +147,6 @@ def package_info(self):
         ]
         if self.settings.os == "Linux":
             self.cpp_info.system_libs.append("pthread")
+        self.cpp_info.names["cmake"] = "absl"
```
Comment (madebr) [inline recipes/abseil/all/conanfile.py:150]:
```suggestion
```
The `cmake` generator is not used to create the `FincAbsl.cmake` or `AbslConfig.cmake` scripts.

Comment (0x8000-0000) [reply recipes/abseil/all/conanfile.py]: Done.

File: recipes/abseil/all/conandata.yml [hunk 2]
```diff
@@ -2,3 +2,6 @@ sources:
   "20200205":
     sha256: 3c554df4909c5c55a6d251f6eadc2c78ff20db5ad4471fd9cbf8085c51b76797
     url: https://github.com/abseil/abseil-cpp/archive/08a7e7bf972c8451855a5022f2faf3d3655db015.tar.gz
+  "20200225.2":
```
Comment (Croydon) [inline recipes/abseil/all/conandata.yml:5]: Abseil did delete their tags in the past, breaking [older versions from our Bincrafters' recipe for it](https://github.com/bincrafters/conan-abseil) and we have no guarantee that they won't do it again
Therefore, I would not be surprised if their delete their old LTS branches on day too

This started this discussion: https://github.com/conan-io/conan-center-index/issues/742

I would feel much more comfortable packaging commits from the master branch

Comment (uilianries) [reply recipes/abseil/all/conandata.yml:5]: @Croydon  Thanks for reporting us! I agree with you, using commit will be better for Abseil.

Comment (madebr) [reply recipes/abseil/all/conandata.yml:5]: Add a comment in the `conandata.yml` file as a reminder for future versions.

File: recipes/abseil/all/conandata.yml [hunk 3]
```diff
@@ -2,3 +2,6 @@ sources:
...
+    url: "https://github.com/abseil/abseil-cpp/archive/20200225.2.tar.gz"
```
Comment (uilianries) [inline recipes/abseil/all/conandata.yml:6]:
```suggestion
    url: "https://github.com/abseil/abseil-cpp/archive/c51510d1d87ebce8615ae1752fd5aca912f6cf4c.tar.gz"
```

File: recipes/abseil/all/conandata.yml (hunk 2 above)
Comment (0x8000-0000) [reply recipes/abseil/all/conandata.yml:8]: Done.

File: recipes/abseil/all/conandata.yml (hunk 3 above)
Comment (0x8000-0000) [reply recipes/abseil/all/conandata.yml]: Done.

**Comment by Croydon** [comment]
For the record: This is a commit which is only on this LTS branch, should they delete the branch than the commit will be purged too, that is why I suggested to go with commits from master as this is probably the only thing the abseil project will not delete/rewrite.

I can see the point in wanting to use the LTS branches, but it might be fragile.

**Comment by danimtb** [comment]
@Croydon I thought the commit was also included in the master branch. So the LTS branches diverge from master? Isn't there an equivalent commit in master we can point this to?

**Comment by jgsogo** [comment]
I really don't know how they organize the repository. There is this LTS branch that started ~2 years ago where they make a commit from time to time (and create a new name for this same branch) and then they have `master` where they are adding lots of commits. If you check the diffs between these two branches they are too far away from each other.

I know they live at HEAD, but are these LTS just patches to the only one release they did in 2018/06?

**Comment by jgsogo** [comment]
Previous existing version of this package (20200205) was a commit to `master` branch: [08a7e7bf972c8451855a5022f2faf3d3655db015](https://github.com/abseil/abseil-cpp/commit/08a7e7bf972c8451855a5022f2faf3d3655db015), I can see that from time to time they _"Export of internal Abseil changes"_ to the `master` branch.  And from time to time they _tag_ a release in the LTS where they create a commit with changes from the `master` branch (https://github.com/abseil/abseil-cpp/commit/b832dce8489ef7b6231384909fd9b68d5a5ff2b7).

Are these commits to LTS the tags we want? 

image

What are those commits to LTS branch? Where do they come from? Are they hotfixes?

**Comment by Croydon** [comment]
Their LTS tags is a new thing which did not exist before (AFAIK), that is why we had not to deal with this before

When they created the current LTS branch: https://github.com/abseil/abseil-cpp/releases/tag/20200225 they mentioned in the commit message `GitOrigin-RevId: 0033c9ea91a52ade7c6b725aa2ef3cbe15463421`, hence I believe that this master commit:  https://github.com/abseil/abseil-cpp/commit/0033c9ea91a52ade7c6b725aa2ef3cbe15463421 is equal to the tag (not git history wise but from the content)

Using 20200225 would also be fully compatible to our own data based version schema 

20200225.1 and 20200225.2 seem to be diverged from master with selective fixes

Unless abseil promises to not delete tags and/or branches anymore, I guess we should treat abseil like a project that does no releases at all and use commit-date based version from the default branch

If we really want to package these tags/diverged branches we probably should consider mirror the source to preserve it

**Comment by Croydon** [comment]
Pinging for @Minimonium's opinion on this matter 😄

**Comment by Croydon** [comment]
By the way, their LTS fixes have references like this: `PiperOrigin-RevId: 298482742`, but I'm not sure what that means. Is this an internal thing only?

**Comment by Minimonium** [comment]
Yes, they dump their internal repo to the git one. This is rather unfortunate since discussions are done internally and it's hard to reach out to googlers that way. 

> Unless abseil promises to not delete tags and/or branches anymore, I guess we should treat abseil like a project that does no releases at all and use commit-date based version from the default branch.

The problem is how their export works. I'm very afraid that they will force push master branch at some point. 😅 

On the point of LTS branches, they're branches for hotfixes and if we assume that they're stable - we 'd need a different from the current one versioning scheme to make packages reproducible. But consider the point of an LTS branch - to provide compatibility. We can't really make use of that with non-semver versions. So I'm sure there is a point for that.

**Comment by 0x8000-0000** [comment]
> (We need to agree on **what is a version for `abseil` package**)

I have opened https://github.com/abseil/abseil-cpp/issues/679 asking for some clarification on their policy.

**Comment by 0x8000-0000** [comment]
> On the point of LTS branches, they're branches for hotfixes and if we assume that they're stable - we 'd need a different from the current one versioning scheme to make packages reproducible. But consider the point of an LTS branch - to provide compatibility. We can't really make use of that with non-semver versions. So I'm sure there is a point for that.

Can you please clarify this comment?

Actually semver has meaning to the upstream project only, on how they consume a dependency. From the packaging perspective, the upstream dependency can name releases by cute mascots - the numbers are completely irrelevant (well, they are handy when used with ranges, but one can clumsily enumerate all compatible versions of the dependencies.)

Or one can consider 20200225 a "major" version - if we must shoehorn this into semver.

So how is 20200225.2 not a good "version number" for this particular tag, if Abseil decides to keep it pointing at c51510d1d87ebce8615ae1752fd5aca912f6cf4c for ever?

**Comment by Minimonium** [comment]
@0x8000-0000 The comment is not specifically about the name of this branch. It's about the Conan special handling of Semver versions. Unless a package uses Semver - consumers can't make use of the backward-compatible nature of LTS branches. Every LTS-patch is a completely separate version, similar to commits in the master branch, every changed version is a fully incompatible configuration.

**Comment by 0x8000-0000** [comment]
> @0x8000-0000 The comment is not specifically about the name of this branch. It's about the Conan special handling of Semver versions. Unless a package uses Semver - consumers can't make use of the backward-compatible nature of LTS branches. Every LTS-patch is a completely separate version, similar to commits in the master branch, every changed version is a fully incompatible configuration.

@Minimonium 

Actually, if they would name the first tag 20200225.0, we would have a basis for Semver, where the origin is the date they cut the branch.

If we run
```
$ git log --pretty=oneline -n 7 20200225.2
```
we see that the tags are "lined" up on the release branches.

Further, if we look for a commit on master around February 25th, we find 0033c9ea91a52ade7c6b725aa2ef3cbe15463421

```
$ git diff --stat 20200225.2..0033c9ea91a52ade7c6b725aa2ef3cbe15463421
 CMakeLists.txt              | 11 +++++++++++
 absl/base/options.h         |  4 ++--
 absl/status/status.cc       |  8 --------
 absl/status/status_test.cc  | 57 ---------------------------------------------------------
 absl/strings/CMakeLists.txt |  1 -
 5 files changed, 13 insertions(+), 68 deletions(-)
```

So at least in practice, assuming these are stable, they "look" like releases. I suppose since they make up the "version numbers" we can make up our "version numbers" as long as the bijection holds and it is clear to a human that if they want Conan package for Abseil 20200225.2 they need to use "abseil/2020.0225.2" .

**Comment by Minimonium** [comment]
https://github.com/abseil/abseil-cpp/issues/679 They guarantee that hashes/tags/LTS branches stay stable. I also double-checked the bincrafters recipe commit history and it seems that indeed for quite a long time package releases were tracked by the commit hash id, instead of a tag. Now I'm now sure from where I've got an idea that tags went missing. 🤔 

Anyway, that means that it'd be a good idea to mirror upstream repository tags while making cuts for master-commits on-demand.

**Comment by 0x8000-0000** [comment]
> Anyway, that means that it'd be a good idea to mirror upstream repository tags while making cuts for master-commits on-demand.

I will update the recipe later tonight and resubmit.

**Comment by 0x8000-0000** [comment]
Reverted to using the tag reference.

**Comment by 0x8000-0000** [comment]
Ping: @madebr @uilianries @danimtb @SSE4

**Pull request merged**